### PR TITLE
[GC] Add struct.new and start to test interesting execution

### PR DESCRIPTION
**Pull request by kripken**
With struct.new read/write support, we can start to do interesting
things! This adds a test of creating a struct and seeing that references
behave like references, that is, if we write to the value X refers to, and
if Y refers to the same thing, when reading from Y's value we see the
change as well.

The test is run through all of `-O1`, which uncovered a minor issue in
Precompute: We can't try to precompute a reference type, as we can't
replace a reference with a value.

Note btw that the test shows the optimizer properly running
CoalesceLocals on reference types, merging two locals.

### CODE REVIEW

**Review by aheejin — APPROVED** [review]
Nice!! 🎉

File: src/wasm/wasm-validator.cpp [hunk 1]
```diff
@@ -2238,7 +2238,33 @@ void FunctionValidator::visitStructNew(StructNew* curr) {
   shouldBeTrue(getModule()->features.hasGC(),
                curr,
                "struct.new requires gc to be enabled");
-  WASM_UNREACHABLE("TODO (gc): struct.new");
+  if (curr->type != Type::unreachable) {
+    if (shouldBeTrue(
+          curr->rtt->type.isRtt(), curr, "struct.new rtt must be rtt")) {
+      auto heapType = curr->rtt->type.getHeapType();
+      if (shouldBeTrue(
+            heapType.isStruct(), curr, "struct.new heap type must be struct")) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/wasm/wasm-validator.cpp:2246]: Yeah, that might be clearer, good point. I'll refactor.

File: src/wasm/wasm-validator.cpp [hunk 2]
```diff
@@ -2238,7 +2238,33 @@ void FunctionValidator::visitStructNew(StructNew* curr) {
...
+        if (curr->isWithDefault()) {
+          // All the fields must be defaultable.
+          for (auto* operand : curr->operands) {
+            shouldBeFalse(
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/wasm/wasm-validator.cpp:2250]: Oh right, thanks! Yes, what I had was wrong...

**Comment by tlively** [comment]
@kripken what's the plan for differentiating between the `_with_rtt` and non-`_with_rtt` variants of instructions? It looks like this PR does not allow `StructNew::rtt` to be null?

**Comment by kripken** [comment]
@tlively I think there is no plan for that - the only thing in the [wasm binary format doc](https://docs.google.com/document/d/1DklC3qVuOdLHSXB5UXghM_syCh-4cMinQ50ICiXnK3Q/edit#heading=h.2p8ajda4s9l1) is the `_with_rtt` variants. That is, it seems like the non-rtt variants have been removed?

**Pull request merged**